src/binary_tree.rs

```diff
@@ -0,0 +1,203 @@
+    fn num_children(&self) -> usize {
+        match self {
+            BinaryNode::EmptyChild | BinaryNode::Leaf(_) | BinaryNode::Hash(_) => 0usize,
+            BinaryNode::Branch(_, _) => 16usize,
+        }
+    }
+
+    /// Return the tree root's *ith* child.
+    ///
+    /// The function will return `EmptyChild` instead of `None` if a branch node has no *ith* child.
+    fn ith_child(&self, index: usize) -> Option<&Self> {
+        if index >= self.num_children() {
+            panic!(format!(
+                "Requested child #{}, only have #{}",
+                index,
+                self.num_children()
+            ));
+        }
+
+        match self {
+            BinaryNode::EmptyChild | BinaryNode::Leaf(_) | BinaryNode::Hash(_) => None,
+            BinaryNode::Branch(ref left, ref right) => match index {
+                0 => Some(left),
+                1 => Some(right),
+                _ => Some(&BinaryNode::EmptyChild),
+            },
+        }
+    }
+
+    fn set_ith_child(&mut self, index: usize, child: &Self) {
+        if index >= self.num_children() {
+            panic!(format!(
+                "Requested child #{}, only have #{}",
+                index,
+                self.num_children()
+            ));
+        }
+
+        match self {
+            BinaryNode::Branch(box left, box right) => {
+                if index == 0 {
+                    if left.is_empty() {
+                        *self = BinaryNode::Branch(Box::new(child.clone()), Box::new(right.clone()))
+                    } else {
+                        panic!("Refusing to overwrite child node");
+                    }
+                } else {
+                    if right.is_empty() {
+                        *self = BinaryNode::Branch(Box::new(left.clone()), Box::new(child.clone()))
+                    } else {
+                        panic!("Refusing to overwrite child node");
+                    }
+                }
+            }
+            _ => panic!("Only branch nodes can be set in this implementation."),
+        }
+    }
+
+    fn children(&self) -> NodeChildIterator<Self, Self> {
+        NodeChildIterator {
+            index: 0,
+            key: None,
+            node: &self,
+        }
+    }
+
+    fn insert(&mut self, key: &BinaryKey, value: Vec<u8>) -> Result<(), String> {
+        use BinaryNode::*;
+
+        if key.is_empty() {
+            return Err("Attempted to insert a 0-byte key".to_string());
+        }
+
+        match self {
+            Leaf(_) => {
+                // If we reach this point, we have reached the bottom of a tree
+                // and this should not happen. Report an error.
+                Err("A key is already present".to_string())
+            }
+            Branch(ref mut left, ref mut right) => {
+                if key.is_empty() {
+                    return Err("Key shorter than the tree depth".to_string());
```

Comment:
If the goal is to have a fixed-depth tree, it's not being enforced. I think with the current implementation it's possible to have leaves in different depths